README.md
Isaac Genome Alignment Software
Copyright (c) 2010-2017 Illumina, Inc.
All rights reserved.

This software is provided under the terms and conditions of the
GNU GENERAL PUBLIC LICENSE Version 3

You should have received a copy of the GNU GENERAL PUBLIC LICENSE Version 3
along with this program. If not, see
<https://github.com/illumina/licenses/>.

---

# Isaac aligner

To configure and install the product, see the [src/INSTALL](src/INSTALL).

The component algorithms for the Isaac aligner are intended for developers to re-use and improve them. 
This version is not commercially supported and provided as is under [GNU GENERAL PUBLIC LICENSE Version 3](https://github.com/illumina/licenses). 
Isaac is included in the [HiSeq Analysis Software described here](http://support.illumina.com/sequencing/sequencing_software/hiseq-analysis-software-v2-0.html) which is commercially supported, and in the Isaac apps on the [BaseSpace platform](https://basespace.illumina.com/home/index). For more information, see our [blog post](http://blog.basespace.illumina.com/2013/06/04/introducing-fast-free-alignment-and-variant-calling-with-the-isaac-human-whole-genome-sequencing-app/).

See [src/Changes](src/Changes) for the release history details.

# Quick start guide

The following example command will get you through aligning some example PhiX data. For more examples and details
please see [src/markdown/manual.md](src/markdown/manual.md)

## Process fastq data

Analyze a pair of fastq files and produce bam output.

    $ <isaac>/bin/isaac-align -r <isaac>/share/*/data/examples/PhiX/iGenomes/PhiX/NCBI/1993-04-28/Sequence/Chromosomes/phix.fa -b <isaac>/share/*/data/examples/PhiX/Fastq -f fastq --use-bases-mask y150,y150 -m40



src/Changes
Isaac-04.20.06.19
* SAAC-1352 ubuntu 20.04 compatibility fixes

Isaac-04.18.11.09
* SAAC-1347 Default max read length to allow for 600
* SAAC-1346 Overestimation of splits in a bin due to large coverage spikes

Isaac-04.18.08.29
* SAAC-1345 Offset table uses 64-bit integers regardless of genome length

Isaac-04.18.06.15
* SAAC-1342 Mismatching number of clusters read for r1/r2 from compressed fastq when zcat|wc -l shows number of reads match

Isaac-04.18.05.23
* SAAC-1341 Isaac runs out of file handles for bins with defaults on human with default linux ulimit -n 1024

Isaac-04.18.05.17
* SAAC-1335 Isaac fails to build with clang

Isaac-04.18.05.11
* SAAC-1340 gcc8 compatibility

Isaac-04.18.05.02
* SAAC-1326 Support for references over 4 gigabases

Isaac-04.18.04.28
* SAAC-1339 Memory control failure with splitInfoList_

Isaac-04.18.04.27
* SAAC-1336 Massive overestimation of splits in non-wgs barcoded samples

Isaac-04.18.02.14
* SAAC-1330 Potential race condition potentially leading to proper pair flag instability

Isaac-04.18.02.08
* SAAC-1329 crash on reads over 256 bases long with default seed length

Isaac-04.18.02.02
* SAAC-1328 incorrect split read counting in alignment leads to memory control failure in bam generation

Isaac-04.18.02.01
* SAAC-1327 TLEN in the bam changes on some reads unless --realign-gaps no is used.

Isaac-04.18.01.19
* SAAC-1318 Update Redist to libxml2-2.9.6 and libxslt-1.1.31

Isaac-04.18.01.18
* SAAC-1324 Incorrect split read buffer estimation leads to bam generation out of memory failures

Isaac-04.18.01.07
* SAAC-1279 Running from Bam I/O does not overlap with compute

Isaac-04.17.12.19
* SAAC-1322 mergeUnmerged cleanup

Isaac-04.17.12.05
* SAAC-1319 Best match candidate collection to fail if the number of best equivalents exceeds repeatThreshold

Isaac-04.17.10.04
* SAAC-1312 readme phix example requires too much RAM
* SAAC-1310 Isaac does not throw errors when dataset has variable read length where --variable-read-length is not set

Isaac-04.17.10.03
* SAAC-1280 Proper pair repeat mate MAPQ upgraded without shadow rescue attempt

Isaac-04.17.09.25
* SAAC-1306 Use second-best candidates for pair building

Isaac-04.17.09.22
* SAAC-1280 Proper pair repeat mate MAPQ upgraded without shadow rescue attempt

Isaac-04.17.09.21
* SAAC-1303 Use third-best seed candidates if there are no second-best seed candidates to improve scoring accuracy

Isaac-04.17.09.20
* SAAC-1304 Stop using normalized gap scores

Isaac-04.17.09.06
* SAAC-1302 When double-checking anomalous pair score the assumed orphan using both seed and rescued candidates
* SAAC-1301 AS tag present in rescued shadow pairs confusing

Isaac-04.17.09.05
* SAAC-1299 Pairing code ignores the limit on maximum number of best pairs
* SAAC-1300 Smith waterman score should be signed
* SAAC-1296 explicit inline for mergeSeedHit brings single tile down to 14 seconds from 16

Isaac-04.17.08.22
* SAAC-1293 Memory control failure when --candidate-matches-max is smaller than --repeat-threshold
* SAAC-1294 --with-dev-stats produces inconsistent counts
* SAAC-1292 compare unclipped positions in validation

Isaac-04.17.08.14
* SAAC-1289 Isaac fails to mark duplicates when inversions have different quality trimming

Isaac-04.17.08.04
* SAAC-1282 --with-dev-stats usability for STE validation

Isaac-04.17.07.21
* SAAC-1274 performance improvements
* SAAC-1275 orphan does not get PE-trimmed when rescuing anomalous repeat mate

Isaac-04.17.07.20
* SAAC-1253 Isaac crashes when using Intel zlib

Isaac-04.17.06.15
* SAAC-1273 specific gapped alignments discard ungapped alignments which sometimes leads to loss of SV evidence

Isaac-04.17.06.13
* SAAC-1258 SV branch performance adjustments
* SAAC-1269 rescued SV pairs are treated as normal, unlike seed-discovered SV pairs
* SAAC-1260 Smith Waterman gaps prevent SV detection
* SAAC-1230 Isaac does not detect introns on 2x63 RNA data
* SAAC-1222 Search for splits only if regular indels are not available
* SAAC-1219 Better test for possiblySemialigned
* SAAC-1204 48 cycles is not enough to detect splits correctly
* SAAC-1130 split reads aligned to repeats with high mapQ scores
* SAAC-1248 Use second best match candidates when looking for splits
* SAAC-1267 best pair collection cleanup

Isaac-04.17.06.09
* SAAC-1272 PeAdapterTrimmer attempts to trim split reads and incorrectly checks for chimera
* SAAC-1271 Inversions cause double counting of soft clipped bases for logProbability

Isaac-04.17.05.18
* SAAC-1262 Account for chimera when computing mapq of pair with splits
* SAAC-1263 gap realigner does not update start position when realigning against deletion on single-ended data

Isaac-04.17.05.11
* SAAC-1254 Oversized rescued pairs are marked as proper

Isaac-04.17.05.08
* SAAC-1254 Oversized rescued pairs are marked as proper

Isaac-04.17.05.05
* SAAC-1243 filter gap realignment candidate gaps by mapq

Isaac-04.17.04.08
* SAAC-1234 SplitReadAligner ignores good split candidates when they don't improve smith waterman score of the constituent

Isaac-04.17.04.07
* SAAC-1238 gap realigner ignores deletions over --split-gap-length

Isaac-04.17.04.06
* SAAC-1229 --realign-vigorously is broken

Isaac-04.17.04.05
* SAAC-1231 support for --response-file command line argument
* SAAC-1233 Process only bins that have some data in bam generation
* SAAC-1221 crash in statistics generation when no data is present in a flowcell

Isaac-04.17.03.01
* SAAC-1209 Up to threads-1 clusters at the end of a tile don't get transposed due to incorrect division of work

Isaac-04.17.02.28
* SAAC-1203 CIGAR buffer overflow when checking for split alignments with --candidate-matches-max 80000
* SAAC-1202 Undersized pairs should be considered proper

Isaac-04.17.02.20
* SAAC-1184 preserve original CIGAR in read names when compiled --with-dev-stats
* SAAC-1192 produce OP tag in bam
* SAAC-1199 Single ended alignment fails when trying to store unaligned records in temp files

Isaac-04.17.02.06

src/c++/lib/common/md5.cpp
/**
 ** Isaac Genome Alignment Software
 ** Copyright (c) 2010-2017 Illumina, Inc.
 ** All rights reserved.
 **
 ** This software is provided under the terms and conditions of the
 ** GNU GENERAL PUBLIC LICENSE Version 3
 **
 ** You should have received a copy of the GNU GENERAL PUBLIC LICENSE Version 3
 ** along with this program. If not, see
 ** <https://github.com/illumina/licenses/>.
 **
 ** \file md5.cpp
 **
 ** copied from the "RSA Data Security, Inc. MD5 Message-Digest Algorithm"
 **
 ** \author Roman Petrovski
 **/

/* md5.c - RSA Data Security, Inc., MD5 Message-Digest Algorithm */

/* NOTE: Numerous changes have been made; the following notice is
ncluded to satisfy legal requirements.

Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
*/

#include <memory.h>
#include "common/md5.hh"

#include <stdint.h>

void MD5Open(MD5Context *md5)
{
  md5->count[0] = md5->count[1] = 0;
  /* Load magic initialization constants.*/
  md5->state[0] = 0x67452301;
  md5->state[1] = 0xefcdab89;
  md5->state[2] = 0x98badcfe;
  md5->state[3] = 0x10325476;
}

/* Constants for MD5Transform routine. */

#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

/* F, G, H and I are basic MD5 functions. */

#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits. */

#define ROTATE_LEFT(x, n) (((x) << (n)) | (((uint32_t)x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
   Rotation is separate from addition to prevent recomputation.
*/

#define FF(a, b, c, d, x, s, ac) { \
 (a) += F((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT((a), (s)); \
 (a) += (b); \
  }


/* MD5 basic transformation. Transforms state based on block. */

static void MD5Transform(UINT4 state[4], const unsigned char block[64])
{
  UINT4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];
  /* Move contents of block to x, putting bytes in little-endian order. */
  #ifdef LITTLE_ENDIAN
    memcpy(x, block, 64);
  #else
  {
    unsigned int i, j;
    for (i = j = 0; i < 16; i++, j+= 4)
    {
      x[i] = (UINT4) block[j] | (UINT4) block[j+1] << 8 |
        (UINT4) block[j+2] << 16 | (UINT4) block[j+3] << 24;
    }
  }
  #endif
  /* Round 1 */
  FF(a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF(d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF(c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF(b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF(a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF(d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF(c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF(b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF(a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF(d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF(c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF(b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF(a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF(d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF(c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF(b, c, d, a, x[15], S14, 0x49b40821); /* 16 */
  /* Round 2 */
  GG(a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG(d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG(c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG(b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG(a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG(d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG(c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG(b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG(a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG(d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG(c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
  GG(b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG(a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG(d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG(c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG(b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */
  /* Round 3 */
  HH(a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH(d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH(c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH(b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH(a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH(d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH(c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH(b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH(a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH(d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH(c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH(b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH(a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH(d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH(c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH(b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */
  /* Round 4 */
  II(a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II(d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II(c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II(b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II(a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II(d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II(c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II(b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II(a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II(d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II(c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II(b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II(a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II(d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II(c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II(b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */
  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;
  /* Zeroize sensitive information. */
  memset(x, 0, sizeof(x));
}

void MD5Digest(MD5Context *md5, const void *input, unsigned int inputLen)
{
  unsigned int i, index, partLen;
  /* Compute number of bytes mod 64 */
  index = (unsigned int)((md5->count[0] >> 3) & 0x3F);
  /* Update number of bits */
  if ((md5->count[0] += ((UINT4)inputLen << 3)) < ((UINT4)inputLen << 3))
    md5->count[1]++;
  md5->count[1] += ((UINT4)inputLen >> 29);
  partLen = 64 - index;
  /* Transform as many times as possible.*/
  if (inputLen >= partLen)
  {
    memcpy(&md5->buffer[index], input, partLen);
    MD5Transform(md5->state, md5->buffer);
    for (i = partLen; i + 63 < inputLen; i += 64)
    {
      // MD5Transform(md5->state, (const unsigned char *) input + i);
      MD5Transform(md5->state, (const unsigned char*) input + i);
    }
    index = 0;
  }
  else
    i = 0;
  /* Buffer remaining input */
  memcpy(&md5->buffer[index], (char *) input + i, inputLen-i);
}

/* ENCODE packs a 32-bit unsigned integer into 4 bytes in little-endian
   order.
*/

#ifdef LITTLE_ENDIAN
#define ENCODE(p,n) *(UINT4 *)(p) = n
#else
#define ENCODE(p,n) (p)[0]=n,(p)[1]=n>>8,(p)[2]=n>>16,(p)[3]=n>>24
#endif

void MD5Close(MD5Context *md5, unsigned char digest[16])
{
  unsigned char bits[8];
  unsigned int index, padLen;
  static unsigned char PADDING[64] =
  {
    0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
  };
  /* Save number of bits */
  ENCODE(bits, md5->count[0]);
  ENCODE(bits+4, md5->count[1]);
  /* Pad out to 56 mod 64. */
  index = (unsigned int)((md5->count[0] >> 3) & 0x3f);
  padLen = (index < 56) ? (56 - index) : (120 - index);
  MD5Digest(md5, PADDING, padLen);
  /* Append length (before padding) */
  MD5Digest(md5, bits, 8);
  /* Store state in digest */
  ENCODE(digest, md5->state[0]);
  ENCODE(digest+4, md5->state[1]);
  ENCODE(digest+8, md5->state[2]);
  ENCODE(digest+12, md5->state[3]);
  /* Zeroize sensitive information. */
  memset(md5, 0, sizeof(MD5Context));
}
//#include <fstream>
//#include <sstream>
//#include <iostream>
//
//
//#include <openssl/md5.h>
//int main(int argc, char *argv[])
//{
//    {
//        MD5Context md5;
//        MD5Open(&md5);
//        std::ifstream is(argv[1]);
//        char szBuffer[1024];
//        while (is)
//        {
//            is.read(szBuffer, sizeof(szBuffer));
//            MD5Digest(&md5, szBuffer, is.gcount());
//        }
//        unsigned char digest[16];
//        MD5Close(&md5, digest);
//
//        std::stringstream sss;
//        for(unsigned int i = 0; i < sizeof(digest); ++i)
//        {
//            const int val = static_cast<int>(digest[i]);
//            sss << ((val < 16)?("0"):("")) << std::hex << val;
//        }
//        std::cout << sss.str() << std::endl;
//
//    }
//
//    {
//        MD5_CTX md5;
//        MD5_Init(&md5);
//        std::ifstream is(argv[1]);
//        char szBuffer[1024];
//        while (is)
//        {
//            is.read(szBuffer, sizeof(szBuffer));
//            MD5_Update(&md5, szBuffer, is.gcount());
//        }
//        unsigned char digest[16];
//        MD5_Final(digest, &md5);
//
//        std::stringstream sss;
//        for(unsigned int i = 0; i < sizeof(digest); ++i)
//        {
//            const int val = static_cast<int>(digest[i]);
//            sss << ((val < 16)?("0"):("")) << std::hex << val;
//        }
//        std::cout << sss.str() << std::endl;
//    }
//
//
//    return 0;
//}
